nomos-blend/network/src/peer_sets.rs

```diff
@@ -0,0 +1,77 @@
+use std::collections::HashSet;
+
+use libp2p::PeerId;
+
+/// Maintain different categories of peers.
+/// Connected peers may be either healthy or unhealthy.
+/// Also, this keeps track of malicious peers, which are eventually disconnected
+/// but shouldn't be connected again.
+#[derive(Debug)]
+pub(crate) struct PeerSets {
+    negotiated_healthy_peers: HashSet<PeerId>,
+    negotiated_unhealthy_peers: HashSet<PeerId>,
+    malicious_peers: HashSet<PeerId>,
+}
+
+impl PeerSets {
+    pub(crate) fn new() -> Self {
+        Self {
+            negotiated_healthy_peers: HashSet::new(),
+            negotiated_unhealthy_peers: HashSet::new(),
+            malicious_peers: HashSet::new(),
+        }
+    }
+
+    /// Return the number of negotiated peers, including healthy and unhealthy peers.
+    pub(crate) fn num_negotiated_peers(&self) -> usize {
+        self.negotiated_healthy_peers.len() + self.negotiated_unhealthy_peers.len()
+    }
+
+    /// Return the number of negotiated healthy peers.
+    pub(crate) fn num_negotiated_healthy_peers(&self) -> usize {
+        self.negotiated_healthy_peers.len()
+    }
+
+    /// Check if the peer has been detected as malicious.
+    pub(crate) fn is_malicious(&self, peer_id: &PeerId) -> bool {
+        self.malicious_peers.contains(peer_id)
+    }
+
+    // Return all negotiated peers, including healthy and unhealthy peers.
+    pub(crate) fn negotiated_peers(&self) -> impl Iterator<Item = &PeerId> {
+        self.negotiated_healthy_peers
+            .iter()
+            .chain(&self.negotiated_unhealthy_peers)
+    }
+
+    /// Return all peers that are negotiated or malicious.
+    pub(crate) fn negotiated_or_malicious_peers(&self) -> impl Iterator<Item = &PeerId> {
+        self.negotiated_peers().chain(&self.malicious_peers)
+    }
+
+    /// Add a new healthy peer.
+    /// If the peer was previously unhealthy, it is removed from the unhealthy set.
+    pub(crate) fn add_healthy(&mut self, peer_id: PeerId) {
+        self.negotiated_unhealthy_peers.remove(&peer_id);
+        self.negotiated_healthy_peers.insert(peer_id);
+    }
+
+    /// Move the peer from the healthy set to the unhealthy set.
+    pub(crate) fn set_unhealthy(&mut self, peer_id: PeerId) {
+        self.negotiated_healthy_peers.remove(&peer_id);
+        self.negotiated_unhealthy_peers.insert(peer_id);
+    }
+
+    /// Move the peer from the healthy or unhealthy set to the malicious set.
+    pub(crate) fn set_malicious(&mut self, peer_id: PeerId) {
+        self.negotiated_healthy_peers.remove(&peer_id);
+        self.negotiated_unhealthy_peers.remove(&peer_id);
+        self.malicious_peers.insert(peer_id);
+    }
```

Comment:
If when setting a peer as malicious they are removed from healthy and unhealthy; shouldn't it happen the other way around? E.g.: Shouldn't `set_unhealthy` remove from `malicious_peers`?

Reply:
Ah, good point. Actually, in the current spec, any peer who has been set as malicious can't be removed from the malicious set. But, yeah, as you said, we should handle the case when the peer is in the malicious set when `set_unhealthy` is called. I will re-think about this state machine and refine it.

Reply:
I removed this `peer_sets.rs` because I found and use [`libp2p-allow-block-list`](https://docs.rs/libp2p-allow-block-list/latest/libp2p_allow_block_list/) and [`libp2p-connection-limits`](https://docs.rs/libp2p-connection-limits/latest/libp2p_connection_limits/) in the end. 4680e946b72d52db6c09695b3f7e8cda67f3c4fc
They simplify this PR a lot. Instead of maintaining the `malicious_peers` set, now we add malicious peers to the libp2p block list, so that the connections from them or to them can be denied / closed.

Reply:
I'll go over it again, no worries :)